libs/execution_context/ebpf_program.c

```diff
@@ -50,44 +51,62 @@ typedef struct _ebpf_program
+    ebpf_helper_function_addresses_t* helper_function_addresses = NULL;
 
-    if (program->program_information_provider_dispatch_table != NULL) {
-        size_t function_count = (program->program_information_provider_dispatch_table->size -
-                                 EBPF_OFFSET_OF(ebpf_extension_dispatch_table_t, function)) /
-                                sizeof(program->program_information_provider_dispatch_table->function[0]);
+    old_provider_data = program->program_information_data;
+    program->program_information_data = NULL;
 
-        if (provider_dispatch_table == NULL) {
-            program->program_invalidated = true;
-            return;
+    if (old_provider_data != NULL) {
+        ebpf_program_data_t* old_program_data = (ebpf_program_data_t*)old_provider_data->data;
+        if (old_program_data != NULL) {
+            old_helper_function_addresses = old_program_data->helper_function_addresses;
         }
+    }
 
-        if (!program->trampoline_table) {
-            return_value = ebpf_allocate_trampoline_table(function_count, &program->trampoline_table);
-            if (return_value != EBPF_SUCCESS) {
-                program->program_invalidated = true;
-                return;
-            }
+    if (provider_data != NULL) {
+        ebpf_program_data_t* program_data = (ebpf_program_data_t*)provider_data->data;
+        if (program_data != NULL) {
+            helper_function_addresses = program_data->helper_function_addresses;
         }
+    }
 
-        return_value = ebpf_update_trampoline_table(
-            program->trampoline_table, program->program_information_provider_dispatch_table);
+    if (helper_function_addresses == NULL) {
+        if (old_helper_function_addresses != NULL)
+            // This happens when the extension is unloading.
+            goto Exit;
+    } else {
+        size_t function_count = helper_function_addresses->helper_function_count;
 
-        if (return_value != EBPF_SUCCESS) {
-            program->program_invalidated = true;
-            return;
+        if (old_helper_function_addresses != NULL) {
+            if (old_helper_function_addresses->helper_function_count != function_count)
+                // A program information provider cannot modify helper function count;
+                goto Exit;
         }
+
+        if (!program->trampoline_table) {
+            return_value = ebpf_allocate_trampoline_table(function_count, &program->trampoline_table);
+            if (return_value != EBPF_SUCCESS)
+                goto Exit;
+        }
+
+        return_value = ebpf_update_trampoline_table(program->trampoline_table, helper_function_addresses);
+        if (return_value != EBPF_SUCCESS)
+            goto Exit;
     }
 
-    program->program_information_provider_dispatch_table = provider_dispatch_table;
     program->program_information_binding_context = provider_binding_context;
     program->program_information_data = provider_data;
```

Comment:
Should we update the program_information_data here? It is possible extension may have changed the program context information.

Reply:
I have made some changes to ensure upon provider reload helper function count is not changed. #313 created to track other issues.